WcaOnRails/app/models/competition.rb

```diff
@@ -458,11 +465,82 @@ def in_progress?
   end
 
   def is_over?
-    end_date < Date.today
+    !!end_date && end_date < Date.today
+  end
+
+  def results_updated_at
+    Option.from_nilable(results.order('updated_at desc').pluck(:updated_at).first)
+  end
+
+  def result_cache_key(view)
+    [id, view, results_updated_at.map(&:iso8601).unwrap_or("")]
```

Comment:
Instead of encoding "there is no results_updated_at date" by returning `nil` and returning a `DateTime` otherwise, I return `None` and `Some.new(...)`. I'm a big fan of method's return values having the same exact interface. That's not the case with `nil`. So, it's mainly to avoid `nil` creeping in since you can't call the methods you want to call without explicitly _unwrapping_ the value first. Having methods like `map` on `Option` also avoids `if`, but that wasn't the main motivation.

The wikipedia article has some more information: https://en.wikipedia.org/wiki/Option_type and I'm borrowing Rust's names: https://doc.rust-lang.org/std/option/enum.Option.html

If the concept is too alien for anyone, I'm going back to returning `nil` from `results_updated_at` even though I'd consider that broken. :see_no_evil: